### PR TITLE
Account for older GCC versions not supporting __has_builtin

### PR DESCRIPTION
**Pull request by illuhad**

### CODE REVIEW

File: src/runtime/adaptivity_engine.cpp [hunk 1]
```diff
@@ -153,7 +153,17 @@ bool is_likely_invariant_argument(common::db::kernel_entry &kernel_entry,
 }
 
 int determine_ptr_alignment(uint64_t ptrval) {
-#if __has_builtin(__builtin_ctz)
+#if defined(__GNUC__)
```
Comment (fodinabor) [inline src/runtime/adaptivity_engine.cpp]: Did Clang always support `__builtin_clz`?
This only tests that the compiler has some GNU C extensions: https://stackoverflow.com/a/55926503

Comment (illuhad) [reply src/runtime/adaptivity_engine.cpp]: Ah damn... clang at least has `__has_builtin` which is certainly supported for all clang >= 14 which we support

Comment (illuhad) [reply src/runtime/adaptivity_engine.cpp]: Okay, I think it should work now correctly :)

**Review by fodinabor — APPROVED** [review]
LGTM

**Review by tdavidcl — APPROVED** [review]
I confirm that it fixes the issue on ubuntu 20 (+llvm15).

```
root@a386bddada30:/home/docker/AdaptiveCpp/build# gcc --version
gcc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Before :
```
[ 84%] Building CXX object src/runtime/CMakeFiles/acpp-rt.dir/adaptivity_engine.cpp.o
/home/docker/AdaptiveCpp/src/runtime/adaptivity_engine.cpp:156:18: error: missing binary operator before token "("
  156 | #if __has_builtin(__builtin_ctz)
      |                  ^
make[2]: *** [src/runtime/CMakeFiles/acpp-rt.dir/build.make:336: src/runtime/CMakeFiles/acpp-rt.dir/adaptivity_engine.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:719: src/runtime/CMakeFiles/acpp-rt.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```

After:
```
Scanning dependencies of target acpp-rt
[ 66%] Building CXX object src/runtime/CMakeFiles/acpp-rt.dir/adaptivity_engine.cpp.o
[ 67%] Building CXX object src/runtime/CMakeFiles/acpp-rt.dir/generic/async_worker.cpp.o
[ 68%] Building CXX object src/runtime/CMakeFiles/acpp-rt.dir/hw_model/memcpy.cpp.o
[ 69%] Building CXX object src/runtime/CMakeFiles/acpp-rt.dir/serialization/serialization.cpp.o
[ 70%] Linking CXX shared library libacpp-rt.so
[ 88%] Built target acpp-rt
```

**Pull request merged**